kb/2023-08-11/using_netapp_third_party_storage.md

```diff
@@ -0,0 +1,281 @@
+      device names matching a certain pattern, and attempt to set up multipathing on them.
+      Unfortunately, Longhorn's device names follow the same pattern, and
+      will not work correctly if `multipathd` tries to use those devices.
+
+      Therefore the file `/etc/multipath.conf` must be set up on each node
+      so as to prevent `multipathd` from touching any of the devices
+      that Longhorn will use.  Unfortunately, it is not possible to know
+      in advance which device names will be used until the volumes are attached
+      to a VM when the VM is started, or when the volumes are hot-added to a running VM.
+      The recommended method is to "whitelist" the Trident devices using device
+      properties rather than device naming.  The properties to allow are the
+      device vendor and product.  Here is an example of what you'll want in `/etc/multipath.conf`:
+
+       ```text
+       blacklist {
+           device {
+               vendor "!NETAPP"
+               product "!LUN"
+           }
+       }
+       blacklist_exceptions {
+           device {
+               vendor "NETAPP"
+               product "LUN"
+           }
+       }
+       ```
+
+       This is only an example that works if NetApp is the only storage provider in the system
+       for which `multipathd` must be used.  More complex environments will require
+       more complex configuration.
+
+       Explicitly putting that content into `/etc/multipath.conf` will work when you start `multipathd` as described below,
+       but the change in `/etc` will not persist across node reboots.  To solve that problem, you should add another
+       file to `/oem` that will re-generate `/etc/multipath.conf` when the node reboots.  The following example
+       will create the `/etc/multipath.conf` given in the example above, but may need to be modified for your
+       environment if you have a more complex iSCSI configuration:
+
+       ```text
+       stages:
+          initramfs:
+            - name: "Configure multipath blacklist and whitelist"
+              files:
+              - path: /etc/multipath.conf
+                permissions: 0644
+                owner: 0
+                group: 0
+                content: |
+                  blacklist {
+                      device {
+                          vendor "!NETAPP"
+                          product "!LUN"
+                       }
+                   }
+                   blacklist_exceptions {
+                       device {
+                           vendor "NETAPP"
+                           product "LUN"
+                       }
+                   }
+       ```
+
+       Again, this has to be done on every node.
+
+   1. Enable multipathd
+
+    Adding the above files to `/oem` will take effect on the next reboot of the node;
+    `multipathd` can be enabled immediately without rebooting the node
+    using the following commands:
```

Comment:
```suggestion
    Adding the above files to `/oem` will take effect on the next reboot of the node; `multipathd` can be enabled immediately without rebooting the node using the following commands:
```

Reply:
Done.  I only just realized that some of what you changed before was to join multiple lines of text into a single line, which I guess affects how markdown decides where to put the line breaks.  Going back to adjust all of the previous corrections I'm making to match that change.